### PR TITLE
added support for regularizers:

### PR DESCRIPTION
**Pull request by fmohr**
- Regularization Techniques for NNs:
    - SWA
    - Lookahead
    - Snapshot Ensembles
- unified application of data transformation in the PreprocessingWorkflow
- not label-encoding the y anymore at the data loading level but only inside of the workflows, so that for later analysis we still have the original labels available when compiling the confusion matrix automatically create constant predictions (independently of the concrete workflow) if the training data has only one class
- resolved issues in the AUCROC metric when there is only one class available
- fixed bugs in LibSVM and also added proper support for the pre-processing
- fixed bugs in the OOB iteration learning curve of RandomForestWorkflow
- introduced epoch-schedule for XGBoost
- resolved bug in the internal labeling of the neural network workflow
- removed unnecessary complicated invocation logic with optional X_valid, X_test etc. this is now always transmitted, and workflows can just ignore them if they don't want to use them.-

### CODE REVIEW

File: publications/2023-neurips/install/requirements.txt
Comment (Deathn0t) [inline publications/2023-neurips/install/requirements.txt:1]: I think this file is outdated and the `keras-swa` package should be added to the `setup.py` instead.

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py [hunk 1]
```diff
@@ -1,17 +1,18 @@
 import keras
 import numpy as np
 from ConfigSpace import Categorical, ConfigurationSpace, Float, Integer
-from lcdb.scorer import ClassificationScorer
-from lcdb.timer import Timer
-from lcdb.utils import get_schedule
-from lcdb.workflow._base_workflow import BaseWorkflow
-from lcdb.workflow.keras.utils import (
+from ...scorer import ClassificationScorer
```
Comment (Deathn0t) [inline publications/2023-neurips/lcdb/workflow/keras/_dense.py:5]: the relative imports here are inconsistent with the code base. replace by absolute imports `lcdb.scorer` same for other relative imports.

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py [hunk 2]
```diff
@@ -21,17 +22,22 @@
     StandardScaler,
 )
 
+# regularization techniques
+from ._lookahead import Lookahead
```
Comment (Deathn0t) [inline publications/2023-neurips/lcdb/workflow/keras/_dense.py:28]: replace by absolute imports

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py [hunk 3]
```diff
@@ -57,6 +63,22 @@
...
+        ),
+        "lookahead": Categorical(
+            "lookahead", items=[True, False], default=True
+        ),
+        "snapshot_ensembles": Categorical(
+            "snapshot_ensembles", items=[False, True], default=True
+        ),
+        # TODO: The following snapshot parameters should be made conditional and only appear if snapshot_ensembles=True
+        "snapshot_ensembles_period": Integer(
+            "snapshot_ensembles_period", bounds=[2, 100], default=20
```
Comment (Deathn0t) [inline publications/2023-neurips/lcdb/workflow/keras/_dense.py]: other hyperparameter with bounds used tuple instead of list `(2, 100)`

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py [hunk 4]
```diff
@@ -155,19 +177,52 @@ def __init__(
         optimizer="Adam",
         learning_rate=0.001,
         batch_size=32,
-        num_epochs=200,
+        num_epochs=1000,
```
Comment (Deathn0t) [inline publications/2023-neurips/lcdb/workflow/keras/_dense.py:190]: this looks like a lot of epochs

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py [hunk 5]
```diff
@@ -155,19 +177,52 @@ def __init__(
...
+        stochastic_weight_averaging=True,
+        lookahead=True,
+        snapshot_ensembles=True,
+        snapshot_ensembles_period=20,
+        snapshot_ensembles_reset_weights=False,
         shuffle_each_epoch=True,
         transform_real="none",
         transform_cat="onehot",
-        verbose=0,
+        verbose=2,
```
Comment (Deathn0t) [inline publications/2023-neurips/lcdb/workflow/keras/_dense.py:204]: maybe set to 2 for debug? should verbose be activated by default?

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py [hunk 6]
```diff
@@ -337,19 +395,44 @@ def _fit(self, X, y, X_valid, y_valid, X_test, y_test, metadata):
             epoch_schedule=self.epoch_schedule,
         )
 
+        # define callbacks
+        callbacks = [
+            keras.callbacks.TerminateOnNaN(),
+            keras.callbacks.ReduceLROnPlateau(),
+            keras.callbacks.EarlyStopping(patience=100),
```
Comment (Deathn0t) [inline publications/2023-neurips/lcdb/workflow/keras/_dense.py:369]: patience looks very large

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py [hunk 7]
```diff
@@ -358,17 +441,36 @@ def _predict(self, X):
         y_pred = self._transformer_label.inverse_transform(y_pred)
         return y_pred
 
-    def _predict_proba(self, X):
-        X = self.transform(X, y=None, metadata=self.metadata).astype(np.float32)
-        y_pred = self.learner.predict(
-            X, batch_size=min(len(X), self.batch_size), verbose=self.verbose
-        )
-        return y_pred
+    def _predict_proba(self, X, apply_transform=True, use_snapshot_ensemble=None):
```
Comment (Deathn0t) [inline publications/2023-neurips/lcdb/workflow/keras/_dense.py]: why is snapshot ensemble `None` isn't it a boolean? why are we not using the `self.` attribute here?

File: publications/2023-neurips/lcdb/workflow/keras/utils.py [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 import keras
 import numpy as np
+import importlib
```
Comment (Deathn0t) [inline publications/2023-neurips/lcdb/workflow/keras/utils.py]: core packages should come first in the list.

**Review by Deathn0t — APPROVED** [review]

File: publications/2023-neurips/install/requirements.txt
Comment (fmohr) [reply publications/2023-neurips/install/requirements.txt:1]: we do not use `keras-swa` anymore, I remove it.

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py (hunk 1 above)
Comment (fmohr) [reply publications/2023-neurips/lcdb/workflow/keras/_dense.py:5]: what's wrong with relative imports? I mean the tests run, so I presume it works correctly?

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py (hunk 2 above)
Comment (fmohr) [reply publications/2023-neurips/lcdb/workflow/keras/_dense.py:28]: why?

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py (hunk 3 above)
Comment (fmohr) [reply publications/2023-neurips/lcdb/workflow/keras/_dense.py]: ok

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py (hunk 5 above)
Comment (fmohr) [reply publications/2023-neurips/lcdb/workflow/keras/_dense.py:204]: right

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py (hunk 4 above)
Comment (fmohr) [reply publications/2023-neurips/lcdb/workflow/keras/_dense.py:190]: In preliminary experiments, I saw that on some datasets, this does not converge within 200 epochs.

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py (hunk 6 above)
Comment (fmohr) [reply publications/2023-neurips/lcdb/workflow/keras/_dense.py:369]: yeah, I however saw that it does oftentimes significantly improves after 30 or 40 epochs of stagnation.

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py (hunk 7 above)
Comment (fmohr) [reply publications/2023-neurips/lcdb/workflow/keras/_dense.py]: This is a tricky one. This is because when making a normal prediction with `predict_proba`, e.g., during training, the snapshots should not be used. So this is not a property only of the network itself but also of how predictions should be made. Or would you suggest another way of doing this?

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py (hunk 1 above)
Comment (Deathn0t) [reply publications/2023-neurips/lcdb/workflow/keras/_dense.py:5]: Nothing big. I am just saying we did not use the relative import syntax anywhere else in the package. Therefore, the "code-format" seems inconsistent by using them here.

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py (hunk 2 above)
Comment (Deathn0t) [reply publications/2023-neurips/lcdb/workflow/keras/_dense.py:28]: PEP 8: https://peps.python.org/pep-0008/
> Absolute imports are recommended, as they are usually more readable and tend to be better behaved (or at least give better error messages) if the import system is incorrectly configured (such as when a directory inside a package ends up on `sys.path`):

File: publications/2023-neurips/lcdb/workflow/keras/_dense.py (hunk 7 above)
Comment (Deathn0t) [reply publications/2023-neurips/lcdb/workflow/keras/_dense.py]: I am not sure to understand what is "a normal prediction. If snapshot ensemble is used, the resulting model is the ensemble of snapshots? So, I thought the `predict_proba` should be the mixture distribution of ensemble members.
